accepted/0003-interactive-audit-resolver.md

```diff
@@ -0,0 +1,165 @@
+- `.npmrc` is not a good option because it doesn't live in the repository and being a dotfile is easy to miss, so a developer could overlook it affecting the audit. Also using it for the purpose of storing resolutions to project specific issues seems counter-intuitive.
+
+A separate file referred to as `audit-resolve.json` has the benefit of being single purpose, easy to track and audit, easy to version and migrate between versions of `npm` and comfortable for the users to review in git history and pull requests.
+
+**postpone, remove, investigate**
+Other options are helpers for more elaborate actions a developer would take when confronted with an issue they can't or don't want to fix. 
+
+Why is postpone useful at all? It's designed to build a secure development culture where one didn't yet form. Without it, a developer under time pressure would mark an issue as ignored with intention to un-mark it later. 
+While shipping with a known vulnerability is a bad practice, NPM's mission with the community should be to empower people to build more secure products and trust their skill and understanding of their project's particular needs. We should also aspire to help teams introduce more secure workflows effortlessly, so letting a build pass without risking compromising security long-term is a win. 
+
+Remove is only useful as a convenience. Imagine a developer introducing `npm audit` and having to go through tens of issues. If they notice one of the first issues is caused by a dependency they no longer use, instead of remembering to clean it up later, they can choose this option.
+
+Investigate option is there to aid the user and direct them towards fixing the issue upstream.
+This RFC doesn't cover potential future usecases where NPM could do things ranging from linking to resources on mitigation to generating local fixes or providing a marketplace of companies offering services fixing the issue for customers (business potential for NPM).
+
+### Alternatives
+
+Currently the only alternative is to manage the issues manually and keep track of the resolutions using conventions created by a team individually. Ignoring dev dependencies and certain level of severity will be provided for `npm audit` so it lowers the barrier of entry to using it in a CI system. 
+
+Paid alternatives for managing security of a node.js project are available, including Snyk, with focus on providing patches to their customers.
+
+## Implementation
+
+reference implementation https://github.com/npm/cli/pull/10
+
+prototype of a working tool (in production use) https://www.npmjs.com/package/npm-audit-resolver
+
+The implementation is, and should remain, runnable standalone as a separate package with minor wrapping code - useful for testing new features without bundling unfinished work with npm cli versions and therefore node.js
```

Comment:
It would be nice to describe a bit more the implementation over here, an example of the cli UX would be very welcomed, maybe just copy or start from the example I floated over https://github.com/npm/cli/issues/525#issuecomment-573928594


```
$ npm audit resolve list
1325 [high]     handlebars: Prototype Pollution
1324 [high]     handlebars: Arbitrary Code Execution
788  [moderate] js-yaml: Denial of Service

$ npm audit resolve get 788
Status: Pending
Level: Moderate
Type: Denial of Service
Package name: js-yaml
Dependency of: tap [dev]
Path: tap > tap-parser > js-yaml
More info: https://npmjs.com/advisories/788

$ npm audit resolve set 788 --status=Read

$ npm audit resolve list
1325 [high]     handlebars: Prototype Pollution
1324 [high]     handlebars: Arbitrary Code Execution

$ npm audit resolve get 788
Status: Read
Level: Moderate
Type: Denial of Service
Package name: js-yaml
Dependency of: tap [dev]
Path: tap > tap-parser > js-yaml
More info: https://npmjs.com/advisories/788
```

Reply:
I believe I found a different way to tackle this.

I think the first step is for `npm audit` to start supporting the `audit-resolve.json` file and leave creating it to external tools with various options for UI.
I'm leaning towards limiting this RFC to just that functionality and then as audit-resolve file standard is accepted, propose a way to edit it.

I've got a few ideas for the resolve command without interactions, but npm-audit-resolver is more comfortable to use than these. I wouldn't bother to implement it as interactive if I didn't need it ;)

Some ideas for non-interactive that I could develop into more detailed RFCs:
- audit resolve runs fix and marks as fixed, puts the rest in and marks as 'none' so user can edit
- audit resolve lists unresolved items and lets the user resolve all items in a certain way with one command where the filters are advisory_id, package, dev
- an almost interactive one where the --continue flag and alike are used like with git rebasing

feels like we need to chat less asynchronously too ;)

Reply:
for sure it would be nice to catch up, I tried reaching out on twitter but let me know if there's a better place to start a conversation 😊

Reply:
I'm not seeing anything on twitter, let's switch to email for scheduling - naugtur@gmail.com
I'll explain the motivations behind what resolver is now and then we can brainstorm what to do about the UX

Reply:
right, twitter sucks these days... it used to notify people whenever someone starts following them

but no worries, I’ll shoot you an email them 😊